File: mobile-android/app/src/main/java/com/example/p2paudio/protocol/PairingPayloadValidator.kt
package com.example.p2paudio.protocol

import com.example.p2paudio.model.FailureCode
import com.example.p2paudio.model.PairingConfirmPayload
import com.example.p2paudio.model.PairingInitPayload
import com.example.p2paudio.model.SessionFailure
import com.example.p2paudio.model.UdpConfirmPayload
import com.example.p2paudio.model.UdpInitPayload

object PairingPayloadValidator {

    fun validateInit(payload: PairingInitPayload, nowUnixMs: Long): SessionFailure? {
        if (payload.version != "2" || payload.phase != "init") {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Invalid init payload version/phase")
        }
        if (payload.expiresAtUnixMs < nowUnixMs) {
            return SessionFailure(FailureCode.SESSION_EXPIRED, "Init payload expired")
        }
        if (payload.sessionId.isBlank() || payload.offerSdp.isBlank()) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Init payload missing required fields")
        }
        return null
    }

    fun validateConfirm(
        payload: PairingConfirmPayload,
        expectedSessionId: String,
        nowUnixMs: Long
    ): SessionFailure? {
        if (payload.version != "2" || payload.phase != "confirm") {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Invalid confirm payload version/phase")
        }
        if (payload.expiresAtUnixMs < nowUnixMs) {
            return SessionFailure(FailureCode.SESSION_EXPIRED, "Confirm payload expired")
        }
        if (payload.sessionId != expectedSessionId) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Session ID does not match")
        }
        if (payload.answerSdp.isBlank()) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Confirm SDP is empty")
        }
        return null
    }

    fun validateUdpInit(payload: UdpInitPayload, nowUnixMs: Long): SessionFailure? {
        if (payload.version != "2" ||
            payload.phase != "udp_init" ||
            payload.transport != "udp_opus"
        ) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Invalid UDP init payload version/phase")
        }
        if (payload.expiresAtUnixMs < nowUnixMs) {
            return SessionFailure(FailureCode.SESSION_EXPIRED, "UDP init payload expired")
        }
        if (payload.sessionId.isBlank() || payload.senderDeviceName.isBlank()) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "UDP init payload missing required fields")
        }
        return null
    }

    fun validateUdpConfirm(
        payload: UdpConfirmPayload,
        expectedSessionId: String,
        nowUnixMs: Long
    ): SessionFailure? {
        if (payload.version != "2" ||
            payload.phase != "udp_confirm" ||
            payload.transport != "udp_opus"
        ) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Invalid UDP confirm payload version/phase")
        }
        if (payload.expiresAtUnixMs < nowUnixMs) {
            return SessionFailure(FailureCode.SESSION_EXPIRED, "UDP confirm payload expired")
        }
        if (payload.sessionId != expectedSessionId) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "Session ID does not match")
        }
        if (payload.receiverDeviceName.isBlank() ||
            payload.receiverPort <= 0 ||
            payload.receiverPort > 65_535
        ) {
            return SessionFailure(FailureCode.INVALID_PAYLOAD, "UDP confirm payload missing required fields")
        }
        return null
    }
}

File: desktop-windows/src/P2PAudio.Windows.Core/Protocol/PairingPayloadValidator.cs
using P2PAudio.Windows.Core.Models;

namespace P2PAudio.Windows.Core.Protocol;

public static class PairingPayloadValidator
{
    public static SessionFailure? ValidateInit(PairingInitPayload payload, long nowUnixMs)
    {
        if (payload.Version != "2" || payload.Phase != "init")
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Invalid init payload version/phase");
        }
        if (payload.ExpiresAtUnixMs < nowUnixMs)
        {
            return new SessionFailure(FailureCode.SessionExpired, "Init payload expired");
        }
        if (string.IsNullOrWhiteSpace(payload.SessionId) ||
            string.IsNullOrWhiteSpace(payload.SenderPubKeyFingerprint) ||
            string.IsNullOrWhiteSpace(payload.OfferSdp))
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Init payload missing required fields");
        }
        return null;
    }

    public static SessionFailure? ValidateConfirm(
        PairingConfirmPayload payload,
        string expectedSessionId,
        long nowUnixMs
    )
    {
        if (payload.Version != "2" || payload.Phase != "confirm")
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Invalid confirm payload version/phase");
        }
        if (payload.ExpiresAtUnixMs < nowUnixMs)
        {
            return new SessionFailure(FailureCode.SessionExpired, "Confirm payload expired");
        }
        if (!string.Equals(payload.SessionId, expectedSessionId, StringComparison.Ordinal))
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Session ID does not match");
        }
        if (string.IsNullOrWhiteSpace(payload.ReceiverPubKeyFingerprint) ||
            string.IsNullOrWhiteSpace(payload.AnswerSdp))
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Confirm payload missing required fields");
        }
        return null;
    }

    public static SessionFailure? ValidateUdpInit(UdpInitPayload payload, long nowUnixMs)
    {
        if (payload.Version != "2" ||
            payload.Phase != "udp_init" ||
            payload.Transport != "udp_opus")
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Invalid UDP init payload version/phase");
        }
        if (payload.ExpiresAtUnixMs < nowUnixMs)
        {
            return new SessionFailure(FailureCode.SessionExpired, "UDP init payload expired");
        }
        if (string.IsNullOrWhiteSpace(payload.SessionId) ||
            string.IsNullOrWhiteSpace(payload.SenderDeviceName))
        {
            return new SessionFailure(FailureCode.InvalidPayload, "UDP init payload missing required fields");
        }
        return null;
    }

    public static SessionFailure? ValidateUdpConfirm(
        UdpConfirmPayload payload,
        string expectedSessionId,
        long nowUnixMs)
    {
        if (payload.Version != "2" ||
            payload.Phase != "udp_confirm" ||
            payload.Transport != "udp_opus")
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Invalid UDP confirm payload version/phase");
        }
        if (payload.ExpiresAtUnixMs < nowUnixMs)
        {
            return new SessionFailure(FailureCode.SessionExpired, "UDP confirm payload expired");
        }
        if (!string.Equals(payload.SessionId, expectedSessionId, StringComparison.Ordinal))
        {
            return new SessionFailure(FailureCode.InvalidPayload, "Session ID does not match");
        }
        if (string.IsNullOrWhiteSpace(payload.ReceiverDeviceName) ||
            payload.ReceiverPort <= 0 ||
            payload.ReceiverPort > ushort.MaxValue)
        {
            return new SessionFailure(FailureCode.InvalidPayload, "UDP confirm payload missing required fields");
        }
        return null;
    }
}
